subprojects/base-services/src/main/java/org/gradle/internal/work/DefaultWorkerLeaseService.java

```diff
@@ -41,26 +44,35 @@
 import static org.gradle.internal.resources.DefaultResourceLockCoordinationService.*;
 import static org.gradle.internal.resources.ResourceLockState.Disposition.*;
 
-public class DefaultWorkerLeaseService implements WorkerLeaseService {
+public class DefaultWorkerLeaseService implements WorkerLeaseService, ParallelismConfigurationListener {
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultWorkerLeaseService.class);
 
-    private final int maxWorkerCount;
+    private volatile int maxWorkerCount;
     private int counter = 1;
     private final Root root = new Root();
 
     private final ResourceLockCoordinationService coordinationService;
     private final ProjectLockRegistry projectLockRegistry;
     private final WorkerLeaseLockRegistry workerLeaseLockRegistry;
+    private final ParallelExecutionManager parallelExecutionManager;
 
-    public DefaultWorkerLeaseService(ResourceLockCoordinationService coordinationService, boolean parallelEnabled, int maxWorkerCount) {
-        this.maxWorkerCount = maxWorkerCount;
+    public DefaultWorkerLeaseService(ResourceLockCoordinationService coordinationService, ParallelExecutionManager parallelExecutionManager) {
+        this.maxWorkerCount = parallelExecutionManager.getParallelismConfiguration().getMaxWorkerCount();
         this.coordinationService = coordinationService;
-        this.projectLockRegistry = new ProjectLockRegistry(coordinationService, parallelEnabled);
+        this.projectLockRegistry = new ProjectLockRegistry(coordinationService, parallelExecutionManager.getParallelismConfiguration().isParallelProjectExecutionEnabled());
         this.workerLeaseLockRegistry = new WorkerLeaseLockRegistry(coordinationService);
+        this.parallelExecutionManager = parallelExecutionManager;
+        parallelExecutionManager.addListener(this);
         LOGGER.info("Using {} worker leases.", maxWorkerCount);
     }
 
     @Override
+    public void onConfigurationChange(ParallelismConfiguration parallelismConfiguration) {
+        this.maxWorkerCount = parallelismConfiguration.getMaxWorkerCount();
```

Comment:
If we change this without holding the state lock, we might give out a few more leases than we should be when the config changes. This is probably okay, just wanted to point it out.